dataset/batch_image.py

```diff
@@ -23,391 +22,436 @@
-        """ Rotate each image in the batch at a random angle
+    @classmethod
+    def _flip_(cls, image, mode):
+        """ Flip images
 
         Parameters
-        -----------
-        components : str
-            a component name or names which data should be rotated
-
-        p : float
-            a probability to apply rotate
-            (0. - don't rotate, .5 - rotate half of images, 1 - rotate all images)
+        ----------
+        image : image in numpy format
 
-        angle : tuple
-            an angle range in the form of (min_angle, max_angle), in degrees
+        mode : {'lr', 'ud'}
+            - 'lr' - apply the left/right flip
+            - 'ud' - apply the upside/down flip
         """
-        if np.random.binomial(1, p) > 0:
-            angle = angle or (-45., 45.)
-            _angle = np.random.uniform(*angle)
-            preserve_shape = kwargs.pop('preserve_shape', True)
-            return self._rotate_one(ix, components, _angle, preserve_shape=preserve_shape, **kwargs)
-        image = self.get(ix, components)
-        return image
+        if mode == 'lr':
+            return image[:, ::-1]
+        elif mode == 'ud':
+            return image[::-1]
+
+    @classmethod
+    def _pad_(cls, image, *args, **kwargs):
+        """ pad an image. args and kwargs are passed to np.pad
+        """
+        return np.pad(image, *args, **kwargs)
 
-    @action
-    def flip(self, axis=None):
-        """ Flip images
+    @classmethod
+    def _invert_(cls, image, channels='all'):
+        """ invert channels
 
         Parameters
         ----------
-        axis : {'h', 'v'}
-            'h' for horizontal (left/right) flip
-            'v' for vertical (up/down) flip
-        direction : {'l', 'r', 'u', 'd'}
+        p : str, tuple of ints
+            probabilities of inverting channels.
+            - str - invert all channels
+            - tuple - (first channel, second channel, ...) - invert given channels
         """
-        if axis == 'h':
-            return self.fliplr()
-        elif axis == 'v':
-            return self.flipud()
-        else:
-            raise ValueError("Parameter axis can be 'h' or 'v' only")
+        if channels == 'all':
+            channels = list(range(image.shape[-1]))
+        inv_multiplier = np.zeros(image.shape[-1], dtype=np.float32)
+        inv_multiplier[np.asarray(channels)] = 255
+        return np.abs(np.ones(image.shape, dtype=np.float32)*inv_multiplier - image.astype(np.float32)).astype(np.uint8)
 
+    @classmethod
+    def _salt_and_pepper_(cls, image, mode=0.5, p_pixel=1, salt=255, pepper=0):
+        """ set pixels' intensities to 0 (pepper) or 255 (salt) randomly. Each pixel is chosen
+        uniformly with probability equals p_pixel
 
-class ImagesBatch(BaseImagesBatch):
-    """ Batch class for 2D images
+        Parameters
+        ----------
+        p_pixel : float, callable
```

Comment:
`mode` and `probability per pixel` might be a difficult concept to grasp quickly.

You might use two params (p_salt, p_pepper) instead of (mode, p_pixel)